### PR TITLE
ROE-2374 Add params to who is making filing

### PR DESCRIPTION
**Pull request by mattch1**
### JIRA link

https://companieshouse.atlassian.net/browse/ROE-2374

### Change description

Updated:

- Who is making the filing controller
- Routing
- Navigation
- Unit tests - NB changed test to expect a choice to be checked.

### Work checklist

- [x] Tests added where applicable
- [x] UI changes meet accessibility criteria

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/utils/navigation.ts [hunk 1]
```diff
@@ -137,13 +137,18 @@ export const NAVIGATION: Navigation = {
   [config.PRESENTER_WITH_PARAMS_URL]: {
     currentPage: config.PRESENTER_PAGE,
     previousPage: () => config.OVERSEAS_NAME_URL,
-    nextPage: [config.WHO_IS_MAKING_FILING_URL]
+    nextPage: [config.WHO_IS_MAKING_FILING_WITH_PARAMS_URL]
```
Comment (mwestacott) [inline src/utils/navigation.ts:140]: Not sure this change is needed. A similar one not there on line 130 when we did the presenter/OE name URLs.

File: src/utils/navigation.ts [hunk 2]
```diff
@@ -137,13 +137,18 @@ export const NAVIGATION: Navigation = {
...
   },
   [config.WHO_IS_MAKING_FILING_URL]: {
     currentPage: config.WHO_IS_MAKING_FILING_PAGE,
     previousPage: () => config.PRESENTER_URL,
     nextPage: [config.DUE_DILIGENCE_URL, config.OVERSEAS_ENTITY_DUE_DILIGENCE_URL]
   },
+  [config.WHO_IS_MAKING_FILING_WITH_PARAMS_URL]: {
+    currentPage: config.WHO_IS_MAKING_FILING_PAGE,
+    previousPage: () => config.PRESENTER_WITH_PARAMS_URL,
```
Comment (mwestacott) [inline src/utils/navigation.ts:149]: Same here. Don't think any of these links should be referencing the _URL_ constants. At least not yet.

File: src/routes/index.ts [hunk 3]
```diff
@@ -144,6 +144,8 @@ router.post(config.PRESENTER_WITH_PARAMS_URL, authentication, navigation.hasOver
 
 router.get(config.WHO_IS_MAKING_FILING_URL, authentication, navigation.hasPresenter, whoIsMakingFiling.get);
 router.post(config.WHO_IS_MAKING_FILING_URL, authentication, navigation.hasPresenter, ...validator.whoIsMakingFiling, checkValidations, whoIsMakingFiling.post);
+router.get(config.WHO_IS_MAKING_FILING_WITH_PARAMS_URL, authentication, navigation.hasPresenter, whoIsMakingFiling.get);
```
Comment (mwestacott) [inline src/routes/index.ts:147]: Could you please move this line up, below 145. So we keep the 'gets' together and the 'posts' together. Consistent with what was done for presenter.

File: src/controllers/who.is.making.filing.controller.ts [hunk 4]
```diff
@@ -1,17 +1,33 @@
...
-  OVERSEAS_ENTITY_DUE_DILIGENCE_URL
+  DUE_DILIGENCE_WITH_PARAMS_URL,
+  OVERSEAS_ENTITY_DUE_DILIGENCE_URL,
+  OVERSEAS_ENTITY_DUE_DILIGENCE_WITH_PARAMS_URL
 } from "../config";
 import { getWhoIsFiling, postWhoIsFiling } from "../utils/who.is.making.filing";
 
 export const get = (req: Request, res: Response, next: NextFunction) => {
-  getWhoIsFiling(req, res, next, WHO_IS_MAKING_FILING_PAGE, PRESENTER_URL);
+  let prevPageUrl = PRESENTER_URL;
```
Comment (mwestacott) [inline src/controllers/who.is.making.filing.controller.ts:18]: This is for the back link which we're not doing yet. Please could you remove it. In theory, no changes required in the 'get' function.

File: test/controllers/who.is.making.filing.controller.spec.ts [hunk 5]
```diff
@@ -128,4 +167,51 @@ describe("Who is making filing controller tests", () => {
...
+
+  describe("POST with url params tests", () => {
+    test(`redirect the ${config.OVERSEAS_ENTITY_DUE_DILIGENCE_URL} page when ${WhoIsRegisteringType.SOMEONE_ELSE} is selected`, async () => {
+      const resp = await request(app)
+        .post(config.WHO_IS_MAKING_FILING_WITH_PARAMS_URL)
+        .send({ [WhoIsRegisteringKey]: WhoIsRegisteringType.SOMEONE_ELSE });
+
+      expect(resp.status).toEqual(302);
+      expect(resp.header.location).toEqual(config.OVERSEAS_ENTITY_DUE_DILIGENCE_URL);
+      expect(mockSetExtraData).toHaveBeenCalledTimes(1);
```
Comment (mwestacott) [inline test/controllers/who.is.making.filing.controller.spec.ts:179]: You have missed assertions here that need adding. These will then bring in further test dependencies/imports that you'll need to set up.

Please have another look at the changes we did for the presenter URLs and the unit test changes done for the POST calls:

https://github.com/companieshouse/overseas-entities-web/pull/1099/files#diff-486722c6890797dc94b8e988d01d514539ef8627a512c8ae5a2acc03780d8ffc

File: src/utils/navigation.ts (hunk 1 above)
Comment (mattch1) [reply src/utils/navigation.ts:140]: It does not make a lot of sense to not have this change we would want to navigate to the next page with parameters of we are on one with them. I have tested this locally and we get the parameters in both urls as expected.

Comment (mwestacott) [reply src/utils/navigation.ts:140]: We didn't have it or need it for the presenter change we did together. I think lets copy that for now (as everyone is copying from the same template) and then we will revisit this later, if required. Please add a subtask to the ROE-1989 story if you are concerned.

File: src/utils/navigation.ts (hunk 2 above)
Comment (mattch1) [reply src/utils/navigation.ts:149]: Not sure I understand this comment.

File: src/routes/index.ts (hunk 3 above)
Comment (mattch1) [reply src/routes/index.ts:146]: Moved.

File: src/utils/navigation.ts (hunk 1 above)
Comment (mattch1) [reply src/utils/navigation.ts]: Just changed this back.

File: src/controllers/who.is.making.filing.controller.ts (hunk 4 above)
Comment (mattch1) [reply src/controllers/who.is.making.filing.controller.ts]: Removed.

File: test/controllers/who.is.making.filing.controller.spec.ts (hunk 5 above)
Comment (mattch1) [reply test/controllers/who.is.making.filing.controller.spec.ts:197]: Updated the tests but the post oe due diligence seems to keep finding only the due diligence url for some reason and failing. Commented the line out for the time being, cannot find what it is that's doing this.

File: src/utils/navigation.ts (hunk 2 above)
Comment (mwestacott) [reply src/utils/navigation.ts:149]: Refer to line 139. When we did that we did not reference the 'WITH_PARAMS' URL constant. So let's just keep copying what was done for presenter (for now and unless there are any issues).

File: test/controllers/who.is.making.filing.controller.spec.ts (hunk 5 above)
Comment (mattch1) [reply test/controllers/who.is.making.filing.controller.spec.ts:197]: On line 190 of the test the return value is mocked but seems to be ignored by the test. Added the extra assertion.

File: src/utils/navigation.ts (hunk 2 above)
Comment (mattch1) [reply src/utils/navigation.ts:149]: changed this.

**Review by mwestacott — COMMENTED** [review]
Added some comments. Please also apply to the other new POST test that returns a 302 code.

File: test/controllers/who.is.making.filing.controller.spec.ts [hunk 6]
```diff
@@ -128,4 +183,64 @@ describe("Who is making filing controller tests", () => {
...
+  describe("POST with url params tests", () => {
+    test(`redirect the ${config.OVERSEAS_ENTITY_DUE_DILIGENCE_URL} page when ${WhoIsRegisteringType.SOMEONE_ELSE} is selected`, async () => {
+      mockIsActiveFeature.mockReturnValueOnce(true);
+      mockGetUrlWithParamsToPath.mockReturnValue(config.OVERSEAS_ENTITY_DUE_DILIGENCE_URL);
+      const resp = await request(app)
+        .post(config.WHO_IS_MAKING_FILING_WITH_PARAMS_URL)
+        .send({ [WhoIsRegisteringKey]: WhoIsRegisteringType.SOMEONE_ELSE });
+
+      expect(resp.status).toEqual(302);
+      expect(resp.text).toContain(config.OVERSEAS_ENTITY_DUE_DILIGENCE_URL);
```
[the remark this answers is not in the document]
Comment (mattch1) [reply test/controllers/who.is.making.filing.controller.spec.ts:196]: Ok

File: test/controllers/who.is.making.filing.controller.spec.ts [hunk 7]
```diff
@@ -128,4 +183,64 @@ describe("Who is making filing controller tests", () => {
       expect(resp.text).toContain(SERVICE_UNAVAILABLE);
     });
   });
+
+  describe("POST with url params tests", () => {
+    test(`redirect the ${config.OVERSEAS_ENTITY_DUE_DILIGENCE_URL} page when ${WhoIsRegisteringType.SOMEONE_ELSE} is selected`, async () => {
+      mockIsActiveFeature.mockReturnValueOnce(true);
+      mockGetUrlWithParamsToPath.mockReturnValue(config.OVERSEAS_ENTITY_DUE_DILIGENCE_URL);
```
[the remark this answers is not in the document]
Comment (mattch1) [reply test/controllers/who.is.making.filing.controller.spec.ts:190]: Ok

File: test/controllers/who.is.making.filing.controller.spec.ts [hunk 8]
```diff
@@ -37,10 +43,20 @@ const mockLoggerDebugRequest = logger.debugRequest as jest.Mock;
 const mockGetApplicationData = getApplicationData as jest.Mock;
 const mockSetExtraData = setExtraData as jest.Mock;
 
+const mockIsActiveFeature = isActiveFeature as jest.Mock;
+
+const mockServiceAvailabilityMiddleware = serviceAvailabilityMiddleware as jest.Mock;
+mockServiceAvailabilityMiddleware.mockImplementation((req: Request, res: Response, next: NextFunction) => next() );
+
```
[the remark this answers is not in the document]
Comment (mattch1) [reply test/controllers/who.is.making.filing.controller.spec.ts:50]: Ok

File: test/controllers/who.is.making.filing.controller.spec.ts [hunk 9]
```diff
@@ -37,10 +43,20 @@ const mockLoggerDebugRequest = logger.debugRequest as jest.Mock;
...
+const mockGetUrlWithParamsToPath = getUrlWithParamsToPath as jest.Mock;
+
```
[the remark this answers is not in the document]
Comment (mattch1) [reply test/controllers/who.is.making.filing.controller.spec.ts:52]: Ok

**Review by Zain-Abbas1 — APPROVED** [review]

**Comment by dlloyd1-ch** [comment]
Looks good to me. Have you tested it in Tilt?

**Pull request merged**